osu.Game/Online/Chat/MessageNotifier.cs

```diff
@@ -0,0 +1,234 @@
+        {
+            if (messages == null || !messages.Any())
+                return;
+
+            HandleMessages(messages.First().ChannelId, messages);
+        }
+
+        /// <remarks>
+        /// Resolves the channel id
+        /// </remarks>
+        public void HandleMessages(long channelId, IEnumerable<Message> messages)
+        {
+            var channel = channelManager.JoinedChannels.FirstOrDefault(c => c.Id == channelId);
+
+            if (channel == null)
+            {
+                Logger.Log($"Couldn't resolve channel id {channelId}", LoggingTarget.Information);
+                return;
+            }
+
+            HandleMessages(channel, messages);
+        }
+
+        public void HandleMessages(Channel channel, IEnumerable<Message> messages)
+        {
+            // don't show if the ChatOverlay and the target channel is visible.
+            if (IsActive && channelManager.CurrentChannel.Value == channel)
+                return;
+
+            foreach (var message in messages)
+            {
+                // ignore messages that already have been read
+                if (message.Id < channel.LastReadId)
+                    return;
+
+                // ignore messages from yourself
+                var localUsername = localUser.Value.Username;
+
+                if (message.Sender.Username == localUsername)
+                    continue;
+
+                if (checkForPMs(channel, message))
+                    continue;
+
+                // change output to bool again if another "message processor" is added.
+                checkForMentions(channel, message, localUsername);
+            }
+        }
+
+        private bool checkForPMs(Channel channel, Message message)
+        {
+            if (!notifyOnChat.Value || channel.Type != ChannelType.PM)
+                return false;
+
+            var existingNotification = privateMessageNotifications.FirstOrDefault(n => n.Username == message.Sender.Username);
+
+            if (existingNotification == null)
+            {
+                var notification = new PrivateMessageNotification(message.Sender.Username, channel);
+                notificationOverlay?.Post(notification);
+                privateMessageNotifications.Add(notification);
+            }
+            else
+            {
+                existingNotification.MessageCount++;
+            }
+
+            return true;
+        }
+
+        private void checkForMentions(Channel channel, Message message, string username)
+        {
+            var words = getWords(message.Content);
+
+            if (!notifyOnMention.Value || !anyCaseInsensitive(words, username))
+                return;
+
+            var notification = new MentionNotification(message.Sender.Username, channel);
+            notificationOverlay?.Post(notification);
+        }
+
+        private static IEnumerable<string> getWords(string input) => Regex.Matches(input, @"\w+").Select(c => c.Value);
```

Comment:
Looks like a regex you might want to compile and reuse.

Reply:
What does "compile" mean in the context of a regex?

Reply:
See the [`RegexOptions.Compiled`](https://docs.microsoft.com/en-us/dotnet/api/system.text.regularexpressions.regexoptions?view=netframework-4.8#System_Text_RegularExpressions_RegexOptions_Compiled) flag, and [some more info here](https://docs.microsoft.com/en-us/dotnet/standard/base-types/regular-expression-options?view=netframework-4.8#compiled-regular-expressions). Should probably be worth it over multiple executions. Can always benchmark to be sure.